wrapanapi/systems/openstack.py

```diff
@@ -55,7 +57,450 @@ def _request_timeout_handler(self, url, method, retry_count=0, **kwargs):
             return self.request(url, method, retry_count=retry_count, **kwargs)
 
 
-class OpenstackSystem(WrapanapiAPIBaseVM):
+class OpenstackInstance(Instance):
+    state_map = {
+        'PAUSED': VmState.PAUSED,
+        'ACTIVE': VmState.RUNNING,
+        'SHUTOFF': VmState.STOPPED,
+        'SUSPENDED': VmState.SUSPENDED
+    }
+
+    def __init__(self, system, raw=None, **kwargs):
+        """
+        Constructor for an EC2Instance tied to a specific system.
+
+        Args:
+            system: an EC2System object
+            raw: the raw novaclient Resource object (if already obtained)
+            uuid: unique ID of the instance
+        """
+        super(OpenstackInstance, self).__init__(system, raw, **kwargs)
+        self._uuid = raw.id if raw else kwargs.get('uuid')
+        if not self._uuid:
+            raise ValueError("missing required kwarg: 'uuid'")
+        self._api = self.system.api
+        self._flavor = None
+
+    @property
+    def _identifying_attrs(self):
+        return {'uuid': self._uuid}
+
+    @property
+    def name(self):
+        self.refresh()
+        return self.raw.name
+
+    @property
+    def uuid(self):
+        return self._uuid
+
+    def refresh(self):
+        """
+        Update instance's raw data
+        """
+        try:
+            self.raw = self._api.servers.get(self._uuid)
+        except os_exceptions.NotFound:
+            raise VMInstanceNotFound(self._uuid)
+        return self.raw
+
+    def _get_state(self):
+        self.refresh()
+
+        inst = self.raw
+
+        if inst.status != "ERROR":
+            return self._api_state_to_vmstate(inst.status)
+        if not hasattr(inst, "fault"):
+            raise VMError("Instance {} in error state!".format(self.name))
+        raise VMError("Instance {} error {}: {} | {}".format(
```

Comment:
**REQUIRED:** I don't think we should ```raise``` here, I think we should return ```ERROR```

Reply:
Sure, this came along from the old code. Updated the logic to log the VM fault details and return `VmState.ERROR`